### PR TITLE
i#1349 build drltrace parallel to drstrace : frontend

### PR DESCRIPTION
**Pull request by mxmssh**
Added drltrace frontend tool.
Changed drltrace's documentation.

### CODE REVIEW

**Comment by mxmssh** [comment]
In the next commit, I'm planning to create tests for drltrace.

File: drltrace/CMakeLists.txt [hunk 1]
```diff
@@ -31,48 +31,86 @@
...
+if (WIN32)
+  set_property(TARGET drltracelib PROPERTY COMPILE_DEFINITIONS
+    "${DEFINES_NO_D};RC_IS_DRLTRACELIB")
+else ()
+  set_property(TARGET drltracelib PROPERTY COMPILE_DEFINITIONS ${DEFINES_NO_D})
+endif ()
+
+set_target_properties(drltracelib PROPERTIES
+  RUNTIME_OUTPUT_DIRECTORY${location_suffix} "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}")
+
```
Comment (mxmssh) [inline drltrace/CMakeLists.txt]: Sorry, I forgot set(DynamoRIO_RPATH ON) here. Otherwise drltrace doesn't know know how to find libdynamorio.so.

File: drltrace/drltrace_frontend.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+# define UNICODE
+# define _UNICODE
+#endif
+
+#include "dr_api.h"
+#include "dr_inject.h"
+#include "dr_config.h"
+#include "dr_frontend.h"
+#include "droption.h"
+#undef TESTANY
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Which header above sets TESTANY?  I don't think that any do, right?

File: drltrace/drltrace_frontend.cpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+#define DRLTRACE_ERROR(msg, ...) do { \
+    fprintf(stderr, "ERROR: " msg "\n", ##__VA_ARGS__);    \
+    fflush(stderr); \
+    exit(1); \
+} while (0)
+
+static droption_t<std::string> op_logdir
+(DROPTION_SCOPE_ALL, "logdir", ".", "Log directory to print library call data",
+ "Specify log directory where library call data will be written, in a separate file per \
+process. The default value is \".\" (current dir) If set to \"-\", data for all \
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: suggest: IMHO it's easier to read when splitting using separated strings: then there's no \ and these subsequent lines indent, making it visually easier to see the separate options.

I would also suggest a blank line between options.

See the style here: https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/common/options.cpp

File: drltrace/drltrace_frontend.cpp [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+processes are printed to stderr (warning: this can be slow).");
+static droption_t<bool> op_only_from_app
+(DROPTION_SCOPE_CLIENT, "only_from_app", false, "Reports only library calls from the app",
+ "Only reports library calls from the application itself, as opposed to all calls even \
+from other libraries or within the same library.");
+static droption_t<bool> op_no_follow_children
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Much simpler to be op_follow_children and default to true, right?

File: drltrace/drltrace_frontend.cpp [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+"(overrides the default, which is to trace all children).");
+static droption_t<bool> op_ignore_underscore
+(DROPTION_SCOPE_CLIENT, "ignore_underscore", false, "Ignores library routine names \
+starting with \"_\".", "Ignores library routine names starting with \"_\".");
+static droption_t<bool> op_help
+(DROPTION_SCOPE_FRONTEND, "help", false, "Print this message.", "Print this message");
+static droption_t<bool> op_version
+(DROPTION_SCOPE_FRONTEND, "version", 0, "Print version number.", "Print version number.");
+static droption_t<unsigned int> op_verbose
+(DROPTION_SCOPE_ALL, "verbose", 1, "Change verbosity (default 1).",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: No need for the "(default 1)".  See the usage message: the default is there.

File: drltrace/drltrace_frontend.cpp [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+"Change verbosity (default 1).");
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Ditto.

File: drltrace/drltrace_frontend.cpp [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+static droption_t<std::string> op_ltracelib_ops
+(DROPTION_SCOPE_CLIENT, "ltracelib_ops",
+ DROPTION_FLAG_SWEEP | DROPTION_FLAG_ACCUMULATE | DROPTION_FLAG_INTERNAL,
+ "", "(For internal use: sweeps up drltracelib options)",
+ "This is an internal option that sweeps up other options to pass to the drltracelib.");
+
+/* check that drltracelib.dll, dynamorio.dll and target executable exist */
+static void
+check_input_files(const char *targer_app_full_name, char *dr_root, char *drltrace_path) {
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: "targer"?  You mean "target", right?

File: drltrace/drltrace_frontend.cpp [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    bool result = false;
+
+    /* check that the target applicaton exist */
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: exists

Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: application

File: drltrace/drltrace_frontend.cpp [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    if (targer_app_full_name[0] == '\0')
+        DRLTRACE_ERROR("target application is not specified");
+    if (drfront_access(targer_app_full_name, DRFRONT_READ, &result) != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("cannot find target application at %s", targer_app_full_name);
+    if (!result)
+        DRLTRACE_ERROR("cannot open target application for read at %s",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: style violation: {}

File: drltrace/drltrace_frontend.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+                       targer_app_full_name);
+
+    /* check that dynamorio's root dir exist and accessible */
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: exists and is

File: drltrace/drltrace_frontend.cpp [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+    is_debug = true;
+#endif
+    int errcode;
+    char *process;
+    process_id_t pid;
+    char dr_option[MAX_DR_CMDLINE];
+    dr_option[0] = '\0';
+
+    if (op_no_follow_children.get_value() == true)
+        _snprintf(dr_option, strlen("-no_follow_children") + 1, "-no_follow_children");
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Use dr_snprintf: _snprintf is Windows-only.  I would expect this to not compile on Linux.  Did it?

I see more instances below.

File: drltrace/drltrace_frontend.cpp [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+#define MAX_DR_CMDLINE (MAXIMUM_PATH*6)
+
+#define DRLTRACE_ERROR(msg, ...) do { \
+    fprintf(stderr, "ERROR: " msg "\n", ##__VA_ARGS__);    \
+    fflush(stderr); \
+    exit(1); \
+} while (0)
+
+static droption_t<std::string> op_logdir
+(DROPTION_SCOPE_ALL, "logdir", ".", "Log directory to print library call data",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: I think we have to use drfront_appdata_logdir() for this as . just won't work for sthg like a Dr. Memory package install.  Plus, on Android drfront_access() is sometimes not sufficient to detect accessibility and we need the try-write stuff in the appdata_logdir code.  Could be a separate pull request.

File: drltrace/drltrace_frontend.cpp [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    drfront_status_t sc;
+    char absolute_logdir_path[MAXIMUM_PATH];
+    bool result;
+
+    sc = drfront_get_absolute_path(logdir, absolute_logdir_path,
+                                   BUFFER_SIZE_ELEMENTS(absolute_logdir_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_absolute_path failed, error code = %d\n", sc);
+
+    if (drfront_access(absolute_logdir_path, DRFRONT_WRITE, &result) != DRFRONT_SUCCESS)
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: See comment above: on Android drfront_access() is sometimes not sufficient to detect accessibility and we need the try-write stuff in the appdata_logdir code.  (But also xref i#220 which could change that.)

File: drltrace/drltrace_frontend.cpp [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+_tmain(int argc, const TCHAR *targv[])
+{
+    char drlibpath[MAXIMUM_PATH];
+#ifdef WINDOWS
+    static const char *libname = "drltracelib.dll";
+#else
+    static const char *libname = "libdrltracelib.so";
+#endif
+#ifdef DEBUG
+    sprintf(drlibpath, "debug/%s", libname);
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: No string routines that don't take a max size: use dr_snprintf

File: drltrace/drltrace_frontend.cpp [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+#else
+    sprintf(drlibpath, "release/%s", libname);
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Ditto

File: drltrace/drltrace_frontend.cpp [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+    /* Convert to UTF-8 if necessary */
+    sc = drfront_convert_args((const TCHAR **)targv, &argv, argc);
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("failed to process args, error code = %d\n", sc);
+#endif
+
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, (const char **)argv,
+                                       &parse_err, &last_index) || argc < 2)
+        DRLTRACE_ERROR("Usage error: %s\n Usage:\n%s\n", parse_err.c_str(),
+              droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Normally indent would line up after the (

File: drltrace/drltrace_frontend.cpp [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+
+    target_app_name = argv[last_index];
+    if (target_app_name == NULL) {
+        DRLTRACE_ERROR("Usage error, target application is not specified.\n Usage:\n%s\n",
+                       droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
+    }
+    sc = drfront_get_app_full_path(target_app_name, full_target_app_path,
+                                   BUFFER_SIZE_ELEMENTS(full_target_app_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_app_full_path failed on %s, error code = %d\n",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: style: {}

File: drltrace/drltrace_frontend.cpp [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+                       target_app_name, sc);
+
+    /* get DR's root directory and drltracelib.dll full path */
+    sc = drfront_get_app_full_path(argv[0], full_frontend_path,
+                                   BUFFER_SIZE_ELEMENTS(full_frontend_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_app_full_path failed on %s, error code = %d\n",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: style: {}

File: drltrace/drltrace_frontend.cpp [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+                       argv[0], sc);
+
+    tmp = full_frontend_path + strlen(full_frontend_path) - 1;
+
+    while (*tmp != DIRSEP && *tmp != ALT_DIRSEP && tmp > full_frontend_path)
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: suggest: comment stating assumptions: you assume this binary is in some subdir like bin/?

File: drltrace/drltrace_frontend.cpp [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+              BUFFER_SIZE_ELEMENTS("../dynamorio"),
+              "%s../dynamorio", full_frontend_path);
+
+    sc = drfront_get_absolute_path(tmp_path, full_dr_root_path,
+                                   BUFFER_SIZE_ELEMENTS(tmp_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_absolute_path failed, error code = %d\n", sc);
+
+    _snprintf(full_drlibtrace_path, BUFFER_SIZE_ELEMENTS(full_frontend_path) +
+              BUFFER_SIZE_ELEMENTS(drlibpath) + BUFFER_SIZE_ELEMENTS(".dll"),
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: The max size should be the size of the destination buffer, unless you're truncating the string which does not seem to be the case

File: drltrace/drltrace_frontend.cpp [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+              "%s%s", full_frontend_path, drlibpath);
+    NULL_TERMINATE_BUFFER(full_dr_root_path);
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Best to move this up where the buffer is filled in

File: drltrace/drltrace_frontend.cpp [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+    NULL_TERMINATE_BUFFER(full_drlibtrace_path);
+
+    if (op_logdir.get_value().c_str() != NULL)
+        /* check access to logdir */
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: style: {}

File: drltrace/drltrace_frontend.cpp [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+    configure_application(full_target_app_path, &argv[last_index],
+                          &inject_data, full_dr_root_path, full_drlibtrace_path);
+
+    if (!dr_inject_process_inject(inject_data, false/*!force*/, NULL))
+        DRLTRACE_ERROR("unable to inject");
+
+    if (!dr_inject_process_run(inject_data))
+        DRLTRACE_ERROR("unable to execute target application");
+
+    printf("%s sucessfully started, waiting app for exit\n", full_target_app_path);
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Best to put this inside a verbose check so user can silence it

**Comment by mxmssh** [comment]
I hope now a compilation for Mac would work.

File: drltrace/drltrace.c [hunk 24]
```diff
@@ -332,7 +318,7 @@ options_init(client_id_t id)
                 USAGE_CHECK(res == 1, "invalid -verbose number");
             }
         } else {
-            NOTIFY("UNRECOGNIZED OPTION: \"%s\""NL, token);
+            VNOTIFY(1, "UNRECOGNIZED OPTION: \"%s\""NL, token);
```
Comment (derekbruening) [inline drltrace/drltrace.c]: Shouldn't this be NOTIFY?  It's a fatal error and this will tell the user which option was invalid, right?

File: drltrace/drltrace_frontend.cpp (hunk 2 above)
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: It's best to reply/mark each comment as Done (or with a message explaining why the suggestion is not being adopted) to help ensure that nothing is overlooked.

File: drltrace/drltrace_frontend.cpp [hunk 25]
```diff
@@ -54,28 +54,41 @@
...
 
+#define DRLTRACE_INFO(level, msg, ...) do { \
+    if (op_verbose.get_value() >= level) {\
+        fprintf(stderr, "INFO: " msg "\n", ##__VA_ARGS__);    \
+        fflush(stderr); \
+    }\
+} while (0)
+
 static droption_t<std::string> op_logdir
 (DROPTION_SCOPE_ALL, "logdir", ".", "Log directory to print library call data",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: I don't see an XXX comment to use drfront_appdata_logdir() as "." will fail in several situations (Android root, Windows install into ProgFiles, etc., as already mentioned in earlier comments)?

File: drltrace/drltrace_frontend.cpp [hunk 26]
```diff
@@ -54,28 +54,41 @@
     exit(1); \
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: (Won't let me comment above here, trying to point at line 45)
What about the TESTANY?

File: drltrace/drltrace_frontend.cpp (hunk 4 above)
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Disagreement?

File: drltrace/drltrace_frontend.cpp [hunk 27]
```diff
@@ -136,7 +150,7 @@ configure_application(char *app_name, char **app_argv, void **inject_data,
     dr_option[0] = '\0';
 
     if (op_no_follow_children.get_value() == true)
-        _snprintf(dr_option, strlen("-no_follow_children") + 1, "-no_follow_children");
+        dr_snprintf(dr_option, MAX_DR_CMDLINE, "-no_follow_children");
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Use BUFFER_SIZE_ELEMENTS for safer coding -- then we won't have an overflow if someone changes the decl and forgets to change here.

File: drltrace/drltrace_frontend.cpp (hunk 13 above)
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Worth a XXX comment if not fixing now, about using drfront_appdata

**Comment by derekbruening** [comment]
I think the Appveyor are unrelated (grr, more flaky tests...) but it looks like the Travis Mac failure is still there.

File: drltrace/drltrace_frontend.cpp (hunk 2 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: Yes, I did it for all you comments but unfortunately forgot to click comment button. So I will add them know.

Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: TESTANY is defines in droption.h

File: drltrace/drltrace_frontend.cpp (hunk 13 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: Yes, I'm planning to prepare additional pull request for that,

File: drltrace/drltrace_frontend.cpp (hunk 25 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: ok, will add that.

File: drltrace/CMakeLists.txt [hunk 28]
```diff
@@ -31,48 +31,86 @@
...
+# drltracelib
 
+set(srcs
+    drltrace.c)
+
+set(external_srcs ../framework/drmf_utils.c)
+
+if (WIN32)
+  set(srcs ${srcs} ${PROJECT_SOURCE_DIR}/make/resources.rc)
+  set(DEFINES_NO_D ${DEFINES_NO_D} RC_IS_DRLTRACE)
```
Comment (mxmssh) [inline drltrace/CMakeLists.txt]: done

File: drltrace/CMakeLists.txt [hunk 29]
```diff
@@ -31,48 +31,86 @@
...
-if (NOT DynamoRIO_INTERNAL OR NOT "${CMAKE_GENERATOR}" MATCHES "Ninja")
-  add_custom_command(TARGET drltrace
-    POST_BUILD
-    COMMAND ${CMAKE_COMMAND}
-    ARGS -E echo "Usage: pass to drconfig or drrun: -t drltrace"
-    VERBATIM)
+add_dependencies(drltracelib api_headers)
+
+if ("${CMAKE_GENERATOR}" MATCHES "Visual Studio")
+  # i#1805: to avoid strtol dup symbol errors (libcmt vs ntdll_imports) we just
```
Comment (mxmssh) [inline drltrace/CMakeLists.txt]: done

File: drltrace/drltrace.c [hunk 30]
```diff
@@ -72,13 +72,29 @@
  *   the library entries.
  */
 
+static inline generic_func_t
+cast_to_func(void *p)
```
Comment (mxmssh) [inline drltrace/drltrace.c]: done

File: drltrace/drltrace.c [hunk 31]
```diff
@@ -72,13 +72,29 @@
...
+    return (generic_func_t) p;
+#ifdef WINDOWS
+#  pragma warning(pop)
+#endif
+}
+
 static uint verbose;
 
-#define NOTIFY(level, fmt, ...) do {          \
+#define ALERT(level, fmt, ...) do {          \
```
Comment (mxmssh) [inline drltrace/drltrace.c]: done

File: drltrace/drltrace.c [hunk 32]
```diff
@@ -186,7 +202,7 @@ iterate_exports(const module_data_t *info, bool add)
             }, { /* EXCEPT */
                 func = NULL;
             });
-            NOTIFY(1, "export %s indirected from "PFX" to "PFX NL,
+            NOTIFY("export %s indirected from "PFX" to "PFX NL,
```
Comment (mxmssh) [inline drltrace/drltrace.c]: done

File: drltrace/drltrace.c [hunk 33]
```diff
@@ -197,7 +213,7 @@ iterate_exports(const module_data_t *info, bool add)
                 IF_DEBUG(bool ok =)
                     drwrap_wrap_ex(func, lib_entry, NULL, (void *) sym->name, 0);
                 ASSERT(ok, "wrap request failed");
-                NOTIFY(2, "wrapping export %s!%s @"PFX NL,
+                NOTIFY("wrapping export %s!%s @"PFX NL,
```
Comment (mxmssh) [inline drltrace/drltrace.c]: done

File: drltrace/drltrace_frontend.cpp (hunk 3 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 4 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: I didn't understand that :) If I set op_follow_children to be true by default how can user set it in false ?

File: drltrace/drltrace_frontend.cpp (hunk 5 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 6 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 7 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 8 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done, done.

File: drltrace/drltrace_frontend.cpp (hunk 9 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 10 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 11 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 14 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 15 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 16 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 17 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 18 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 19 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: yes, added comment

File: drltrace/drltrace_frontend.cpp (hunk 20 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 21 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 22 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 23 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

**Comment by mxmssh** [comment]
Sorry, forgot to click comment button after I answered all your previous comments.

**Comment by mxmssh** [comment]
Yes, MAC build again failed. How can I prepare the same environment to test it ?
I see the following error:
Undefined symbols for architecture i386:
	clang: error: linker command failed with exit code 1 (use -v to see invocation)
	make[2]: *** [bin/release/libdrltracelib.dylib] Error 1
	make[1]: *** [drltrace/CMakeFiles/drltracelib.dir/all] Error 2
It's no enough for me to say what happens, so, I would like to prepare the same environment.

File: drltrace/drltrace_frontend.cpp (hunk 13 above)
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: In general, we want to have comments prefixed by XXX (or FIXME for correctness issue) pointing at missing functionality, so it's clear from each patchset what's not there yet.

File: drltrace/drltrace_frontend.cpp (hunk 4 above)
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: -no_ (or other variants) will negate a boolean: see https://github.com/DynamoRIO/dynamorio/blob/master/ext/droption/droption.h#L413

File: drltrace/drltrace_frontend.cpp (hunk 2 above)
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: Why is undefining it necessary?

**Comment by derekbruening** [comment]
> Yes, MAC build again failed. How can I prepare the same environment to test it ?
> I see the following error:
> Undefined symbols for architecture i386:
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [bin/release/libdrltracelib.dylib] Error 1
> make[1]: *** [drltrace/CMakeFiles/drltracelib.dir/all] Error 2
> It's no enough for me to say what happens, so, I would like to prepare the same environment.

To reduce verbosity we turned off -VV and are only running -V, and runsuite_common_post.cmake tries to pull out the error lines, but it looks like the actual missing symbol is not shown there, unfortunately.  The environment is whatever Travis provides: you can check their docs.  If they don't say the clang version you might have to do a pull request that adds -v to get the version.  The same clang version on Linux might reproduce the error (note that the Linux clang run on Travis did not though: maybe different version?), or if you have access to a Mac.  Or you can add a -V back to runsuite_wrapper.pl and do a run to get more info out of Travis.

File: drltrace/drltrace_frontend.cpp (hunk 2 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: To avoid warning.

Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: But from where?  Is it that utils.h also defines it?  And nobody else in drmem repo uses both so we never hit this before? Maybe worth a comment "/* droption.h and utils.h both define this */" or sthg.

Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: utils.h defines that at line 193 (#define TESTANY TEST) and droption.h defines that at line 47 (#define TESTANY(mask, var) (((mask) & (var)) != 0))

File: drltrace/drltrace_frontend.cpp (hunk 25 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 27 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace.c (hunk 24 above)
Comment (mxmssh) [inline drltrace/drltrace.c]: done

File: drltrace/drltrace_frontend.cpp (hunk 13 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done.

File: drltrace/drltrace_frontend.cpp (hunk 4 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

**Comment by mxmssh** [comment]
I've prepared commit with -VV to get out more info from Travis.

**Comment by mxmssh** [comment]
Ok, much better
Undefined symbols for architecture i386:
  "_dr_symbol_export_iterator_hasnext", referenced from:
      _iterate_exports in drltrace.c.o
  "_dr_symbol_export_iterator_next", referenced from:
      _iterate_exports in drltrace.c.o
  "_dr_symbol_export_iterator_start", referenced from:
      _iterate_exports in drltrace.c.o
  "_dr_symbol_export_iterator_stop", referenced from:
      _iterate_exports in drltrace.c.o

**Comment by derekbruening** [comment]
Are those details not in the build xml file?  If you see a simple tweak to runsuite_common_post.cmake to get the details please submit it.  -VV is something of a pain in general as the data is too much for the online viewers and you have to download the raw log.  Plus for Appveyor the online viewer is really slow with that much data.

**Comment by mxmssh** [comment]
Finally I managed to build the same environment and reproduce the error. 
Fixed CMake file to be able to build on Mac. Disabled -VV option.

File: drltrace/drltrace_frontend.cpp [hunk 34]
```diff
@@ -0,0 +1,340 @@
...
+ "process.  The default value is \".\" (current dir).  If set to \"-\", data for all "
+ "processes are printed to stderr (warning: this can be slow).");
+
+static droption_t<bool> op_only_from_app
+(DROPTION_SCOPE_CLIENT, "only_from_app", false, "Reports only library calls from the app",
+ "Only reports library calls from the application itself, as opposed to all calls even "
+ "from other libraries or within the same library.");
+
+static droption_t<bool> op_follow_children
+(DROPTION_SCOPE_FRONTEND, "follow_children", true, "Do not trace child processes",
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: The description is still negated.

File: drltrace/drltrace_frontend.cpp [hunk 35]
```diff
@@ -0,0 +1,340 @@
...
+/* check that drltracelib.dll, dynamorio.dll and target executable exist */
+static void
+check_input_files(const char *target_app_full_name, char *dr_root, char *drltrace_path) {
+    bool result = false;
+
+    /* check that the target application exists */
+    if (target_app_full_name[0] == '\0')
+        DRLTRACE_ERROR("target application is not specified");
+
+    /* FIXME: We need to use drfront_appdata_logdir to handle specific situations when
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: add i# (so "FIXME i#xxx: We ...")

**Comment by derekbruening** [comment]
Appveyor failure looks like #1712

File: drltrace/drltrace_frontend.cpp (hunk 34 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp (hunk 35 above)
Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: done

File: drltrace/drltrace_frontend.cpp [hunk 36]
```diff
@@ -73,7 +73,7 @@ static droption_t<bool> op_only_from_app
  "from other libraries or within the same library.");
 
 static droption_t<bool> op_follow_children
-(DROPTION_SCOPE_FRONTEND, "follow_children", true, "Do not trace child processes",
+(DROPTION_SCOPE_FRONTEND, "follow_children", true, "Trace child processes",
  "(overrides the default, which is to trace all children).");
```
Comment (derekbruening) [inline drltrace/drltrace_frontend.cpp]: The long description is still weird, please consider updating it at some point.

Comment (mxmssh) [inline drltrace/drltrace_frontend.cpp]: Right :) I am so inattentive....

**Comment by derekbruening** [comment]
Appveyor failure is #1712 so I will override.  Also note it should be "typo" not "tipo" -- I will change.

**Comment by mxmssh** [comment]
Ok, thank you, I will update it in the next pull request.
